pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -240,109 +241,220 @@ func substitueYamlVariables(baseYaml []byte, yamlSubstitutions map[string]string
 
-	labels := map[string]string{
-		"component": applicationName,
-	}
+			_, gvk, err := yamlDecoder.Decode([]byte(deployYaml), nil, deploymentManifest)
+			if err != nil {
+				return errors.Wrap(err, "Failed to decode the manifest yaml")
+			}
 
-	manifestLabels := deploymentManifest.GetLabels()
-	if manifestLabels != nil {
-		for key, value := range labels {
-			manifestLabels[key] = value
-		}
-		deploymentManifest.SetLabels(manifestLabels)
-	} else {
-		deploymentManifest.SetLabels(labels)
-	}
+			deployJSON, err := deploymentManifest.MarshalJSON()
+			if err != nil {
+				return errors.Wrap(err, "Deployment manifest is invalid")
+			}
+			klog.V(3).Infof("Deploy manifest:\n\n%s\n", deployJSON)
 
-	// TODO: Determine why using a.Client.DynamicClient doesnt work
-	// Need to create my own client in order to get the dynamic parts working
-	myclient, err := dynamic.NewForConfig(a.Client.KubeClientConfig)
-	if err != nil {
-		return err
-	}
+			gvr := schema.GroupVersionResource{Group: gvk.Group, Version: gvk.Version, Resource: strings.ToLower(gvk.Kind + "s")}
+			klog.V(3).Infof("Manifest type: %s", gvr.String())
+
+			labels := map[string]string{
+				"component": applicationName,
+			}
 
-	// Check to see whether deployed resource already exists. If not, create else update
-	// Get?
-	instanceFound := false
-	list, err := myclient.Resource(gvr).Namespace(namespace).List(metav1.ListOptions{})
-	if list != nil && len(list.Items) > 0 {
-		for _, item := range list.Items {
-			klog.V(3).Infof("Found %s %s with resourceVersion: %s.\n", gvk.Kind, item.GetName(), item.GetResourceVersion())
-			if item.GetName() == applicationName {
-				deploymentManifest.SetResourceVersion(item.GetResourceVersion())
-				instanceFound = true
+			manifestLabels := deploymentManifest.GetLabels()
+			if manifestLabels != nil {
+				for key, value := range labels {
+					manifestLabels[key] = value
+				}
+				deploymentManifest.SetLabels(manifestLabels)
+			} else {
+				deploymentManifest.SetLabels(labels)
+			}
+
+			// Check to see whether deployed resource already exists. If not, create else update
+			instanceFound := false
+			list, err := a.Client.DynamicClient.Resource(gvr).Namespace(namespace).List(metav1.ListOptions{})
+			if list != nil && len(list.Items) > 0 {
+				for _, item := range list.Items {
+					if item.GetName() == applicationName {
+						deploymentManifest.SetResourceVersion(item.GetResourceVersion())
+						deploymentManifest.SetAnnotations(item.GetAnnotations())
+						if item.GetKind() == "Service" {
+							currentServiceSpec := item.UnstructuredContent()["spec"].(map[string]interface{})
+							if currentServiceSpec["type"] == "ClusterIP" {
+								newService := deploymentManifest.UnstructuredContent()
+								newService["spec"].(map[string]interface{})["clusterIP"] = currentServiceSpec["clusterIP"]
+								deploymentManifest.SetUnstructuredContent(newService)
+							}
+						}
+						instanceFound = true
```

Comment:
should we `break` from the for here?

Reply:
Yes, but if we do a Get rather than a List it wont be an issue and this code will get refactored. Will fix in new PR.

Reply:
Use of Get had made doing a break unecessary.